### PR TITLE
Fix cert SN checking

### PR DESCRIPTION
**Pull request by pamaury**
The code currently does not handle "short" serial numbers, ie in the rare cases where the SN would fit on say 19 bytes instead of 20. This causes issues in CI. This commits fixes that.

I suspect there is another problem because I can't actually ever see the SN match on reboot so there might be a problem with the encoding (LE 32-bit vs BE 8-bit). This commit does not address this problem.

This is a temporary patch to address #22921

### CODE REVIEW

**Review by vogelpi — APPROVED** [review]
Thanks for looking into this @pamaury . We're seeing these failures for at least 3 M3 hardware PRs now. This most likely will fix those failures.

**Comment by pamaury** [comment]
I need to fix the corresponding unit tests.

File: sw/device/silicon_creator/lib/cert/cert.c [hunk 1]
```diff
@@ -78,9 +82,10 @@ rom_error_t cert_x509_asn1_check_serial_number(
   for (size_t i = 0; i < kCertX509Asn1SerialNumberSizeIn32BitWords; ++i) {
     curr_sn_word = 0;
     for (size_t j = 0; j < sizeof(uint32_t); ++j) {
-      curr_sn_word |=
-          (uint32_t)cert_bytes[sn_bytes_offset + (i * sizeof(uint32_t)) + j]
-          << (8 * j);
+      size_t idx = (i * sizeof(uint32_t)) + j;
+      if (idx < asn1_integer_length) {
+        curr_sn_word |= (uint32_t)cert_bytes[sn_bytes_offset + idx] << (8 * j);
+      }
```
Comment (timothytrippel) [inline sw/device/silicon_creator/lib/cert/cert.c:88]: what happens if the serial number is 19 bytes? is the comparison off by one byte then?

Maybe this would be cleaner if we just copied the serial number bytes out of cert_bytes into a separate fixed size 32-byte buffer with the bytes in LE order and the proper 0 padding in the MSB if needed. The comparison can then be a simple byte-by-byte  check with `expected_sn_bytes`. WDYT?

**Review by timothytrippel — COMMENTED** [review]
Additionally, could you add some unit tests to the cert_unittest file that check the 21 / <20 byte serial number corner cases?

**Comment by timothytrippel** [comment]
Thanks Amaury for the fix!

**Comment by pamaury** [comment]
After discussing with @timothytrippel , it seems the fix will involve more work on the unittest side so I have for now just disabled the certificate validation entirely to fix the CI at least.

**Review by timothytrippel — APPROVED** [review]
Thanks @pamaury . I can submit the proper fix later today.

**Comment by andreaskurth** [comment]
There still seem to be timeouts unfortunately. Here's one log excerpt:
```
==================== Test output for //sw/device/tests:pwrmgr_sleep_wake_5_bug_test_fpga_cw310_sival_rom_ext:
Invoking test: sw/host/tests/chip/pwrmgr/sleep_all_wakeups --rcfile= --logging=info --interface=hyper310 --bitstream=hw/bitstream/universal/splice.bit --bootstrap=sw/device/tests/pwrmgr_sleep_wake_5_bug_test_fpga_cw310_sival_rom_ext.img
[2024-05-02T18:49:46.233Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=28: No such device (it may have been disconnected)
[2024-05-02T18:49:46.233Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=36: No such device (it may have been disconnected)
[2024-05-02T18:49:46.233Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=44: No such device (it may have been disconnected)
[2024-05-02T18:49:46.233Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=117: No such device (it may have been disconnected)
[2024-05-02T18:49:46.233Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=74: No such device (it may have been disconnected)
[2024-05-02T18:49:46.233Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=108: No such device (it may have been disconnected)
[2024-05-02T18:49:46.233Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=105: No such device (it may have been disconnected)
[2024-05-02T18:49:46.233Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=87: No such device (it may have been disconnected)
[2024-05-02T18:49:46.233Z INFO  opentitanlib::util::usb] Could not open device at bus=1 address=39: No such device (it may have been disconnected)
[2024-05-02T18:49:46.262Z INFO  opentitanlib::io::uart] set_flow_control to false
[2024-05-02T18:49:46.262Z INFO  opentitanlib::test_utils::load_bitstream] Loading bitstream: "hw/bitstream/universal/splice.bit"
[2024-05-02T18:49:46.272Z INFO  opentitanlib::util::usr_access] Bitstream file USR_ACCESS value: 0x14fc52b9
[2024-05-02T18:49:46.336Z INFO  opentitanlib::util::rom_detect] Current bitstream: "ROM:14fc52b9\r"
[2024-05-02T18:49:46.336Z INFO  opentitanlib::transport::common::fpga] Already running the correct bitstream.  Skip loading bitstream.
[2024-05-02T18:49:46.337Z INFO  opentitanlib::test_utils::init] load_bitstream: success.
"load_bitstream": true
[2024-05-02T18:49:46.338Z INFO  opentitanlib::bootstrap] Asserting bootstrap pins...
[2024-05-02T18:49:46.342Z INFO  opentitanlib::app] Asserting the reset signal
[2024-05-02T18:49:46.443Z INFO  opentitanlib::app] Clearing the UART RX buffer
[2024-05-02T18:49:46.443Z INFO  opentitanlib::app] Deasserting the reset signal
[2024-05-02T18:49:46.545Z INFO  opentitanlib::bootstrap] Performing bootstrap...
[2024-05-02T18:49:47.289Z INFO  opentitanlib::bootstrap] Releasing bootstrap pins, resetting device...
[2024-05-02T18:49:47.294Z INFO  opentitanlib::app] Asserting the reset signal
[2024-05-02T18:49:47.395Z INFO  opentitanlib::app] Deasserting the reset signal
[2024-05-02T18:49:47.497Z INFO  opentitanlib::test_utils::init] bootstrap: success.
"bootstrap": true
68:Starting test sleep_all_wakeups_test...
[2024-05-02T18:49:47.497Z  console]ROM:14fc52b9
[2024-05-02T18:49:47.497Z  console]bootstrap:1
[2024-05-02T18:49:47.497Z  console]ROM:14fc52b9
[2024-05-02T18:49:47.497Z  console]Starting ROM_EXT
[2024-05-02T18:49:47.968Z  console]Warning: UDS certificate not valid.
[2024-05-02T18:49:48.003Z  console]CDI_0 certificate not valid. Updating it ...
[2024-05-02T18:49:48.075Z  console]CDI_1 certificate not valid. Updating it ...
[2024-05-02T18:49:48.118Z  console]0: 40130000 NAPOT L--- sz=00001000
[2024-05-02T18:49:48.121Z  console]1: 40480000 NAPOT L--- sz=00000400
[2024-05-02T18:49:48.125Z  console]2: 20010400 ----- ---- sz=00000000
[2024-05-02T18:49:48.128Z  console]3: 20017074   TOR -X-R sz=00006c74
[2024-05-02T18:49:48.131Z  console]4: 00000000 ----- ---- sz=00000000
[2024-05-02T18:49:48.134Z  console]5: 00000000 ----- ---- sz=00000000
[2024-05-02T18:49:48.137Z  console]6: 00000000 ----- ---- sz=00000000
[2024-05-02T18:49:48.140Z  console]7: 00000000 ----- ---- sz=00000000
[2024-05-02T18:49:48.143Z  console]8: 00000000 ----- ---- sz=00000000
[2024-05-02T18:49:48.146Z  console]9: 00000000 ----- ---- sz=00000000
[2024-05-02T18:49:48.150Z  console]10: 20000400 ----- ---- sz=00000000
[2024-05-02T18:49:48.153Z  console]11: 2000ac54   TOR -X-R sz=0000a854
[2024-05-02T18:49:48.156Z  console]12: 20000000 NAPOT L--R sz=00100000
[2024-05-02T18:49:48.159Z  console]13: 00010000 NAPOT L--- sz=00001000
[2024-05-02T18:49:48.162Z  console]14: 40000000 NAPOT L-WR sz=10000000
[2024-05-02T18:49:48.166Z  console]15: 10000000 NAPOT L-WR sz=00020000
[2024-05-02T18:49:48.169Z  console]mseccfg = 00000002
[2024-05-02T18:49:48.171Z  console]entry: 0x20010480
[2024-05-02T18:49:48.173Z  console]I00001 ottf_main.c:154] Running sw/device/tests/pwrmgr_sleep_wake_5_bug_test.c
[2024-05-02T18:49:48.180Z  console]I00002 pwrmgr_sleep_wake_5_bug_test.c:112] Wake_status 0x0
[2024-05-02T18:49:48.186Z  console]I00003 pwrmgr_sleep_all_wake_ups_impl.c:226] Issue WFI to enter sleep 5
[2024-05-02T18:49:48.192Z  console]
[2024-05-02T18:49:48.192Z  console]I00004 pwrmgr_sleep_wake_5_bug_test.c:115] Woke up by source 5
[2024-05-02T18:49:48.198Z  console]
[2024-05-02T18:49:48.198Z  console]I00005 pwrmgr_sleep_wake_5_bug_test.c:119] Wake_status 0x20
[2024-05-02T18:49:48.204Z  console]I00006 pwrmgr_sleep_all_wake_ups_impl.c:226] Issue WFI to enter sleep 5
[2024-05-02T18:49:48.210Z  console]
[2024-05-02T18:49:48.222Z  console]ROM:14fc52b9
[2024-05-02T18:49:48.293Z  console]Starting ROM_EXT
[2024-05-02T18:49:48.782Z  console]Warning: UDS certificate not valid.
[2024-05-02T18:49:48.817Z  console]CDI_0 certificate not valid. Updating it ...
-- Test timed out at 2024-05-02 18:50:46 UTC --
```

**Comment by timothytrippel** [comment]
looks like this is causing a CI timeout because now certs are always updating during ROM_EXT execution... Let me close this PR and submit another one that will temporarily fix CI in the meantime, while I role out the permanent fix one side.
